### ISSUE TITLE
Date Bug when time is 20:00-4:00

### ISSUE THREAD

**Issue by Clubfan22** [SemanticMediaWiki/SemanticMediaWiki#2454]
### Setup and configuration

- SMW version: 2.5.1
- MW version: 1.28.2
- PHP version: 5.6.30-0+deb8u1 (fpm-fcgi)
- DB (MySQL etc.): 10.1.22-MariaDB-1~jessie

See [here](http://wiki.teamliquid.net/counterstrike/Special:Version?detail=1)

### Issue
Hi everyone,

we just updated SMW to version 2.5.2 yesterday and everything is working fine so far, except for one thing:

When trying to store a date with the time defined as "20:00-4:00", an error occurs:

```
Unable to interpret the "May 14, 2017 20:00-4:00" input value as valid JD (Julian day) number with "Part of the date is out of bounds." being reported.
```

It happens on both our wiki ( [actual page](http://wiki.teamliquid.net/counterstrike/Esports_Championship_Series/Season_3/North_America/Week_1-4), [test page](http://wiki.teamliquid.net/counterstrike/User:Clubfan/DateTest) ) and the SMW sandbox wiki, where I was able to [recreate the issue](https://sandbox.semantic-mediawiki.org/wiki/Possible_date_bug). What's interesting in my opinion is that other dates which should represent the same instant in time (like 21:00-5:00 or 19:00-3:00) work as intended.

Thank you already for your help!

**Comment by Clubfan22** [comment]
Another staff member of the wiki came up with this, which might explain it:

http://wiki.teamliquid.net/counterstrike/User:ChapatiyaqPTSM/SMW

**Comment by kghbln** [comment]
@Clubfan22 Thanks for reporting. The test pages are particularly useful to see the issue in action.

**Comment by kghbln** [comment]
I personally wonder why we wiki thinks that we act within the ranges of the Julian calendar.

**Comment by FO-nTTaX** [comment]
Whilst looking over more of hour pages, one of hour contributors noticed a patters as can be seen here now: http://wiki.teamliquid.net/counterstrike/User:ChapatiyaqPTSM/SMW#3_hours

**Comment by mwjames** [comment]
> It happens on both our wiki ( actual page, test page ) and the SMW sandbox wiki, where I was able to recreate the issue.

Thanks for the examples.

This issue has given me a bit of a hard time since the normal `var_dump( ... )` wouldn't reveal the issue with the value output always being cloaked as 2457869.5 for the JD value where in fact the computed [0] value for April 25, 2017 20:00-4:00 was 2457869.4999999665.

When converted into a hour, minute, second representation the value turned into 24:00:00 opposed to an expected 00:00:00. In the end, only PHPUnit revealed that we are dealing with 2457869.4999999665 and not 2457869.5.

Examples of yours have now been included as as part of an integration test.

[0] `$this->m_dataitem->getJD() - $timeoffset / 24`

**Comment by kghbln** [comment]
The fix will ship with SMW 2.5.2 next week.
